contracts/NFTVesting/IntervalVestingNftTimeLock.sol

```diff
@@ -0,0 +1,234 @@
+    function nft() public view virtual returns (IERC721) {
+        return _nft;
+    }
+
+    /**
+     * @dev Returns the token ID of the NFT being held.
+     * Returns undefined if the contract is not holding an NFT.
+     */
+    function tokenId() public view virtual returns (uint256) {
+        return _tokenId;
+    }
+
+    /**
+     * @dev Returns the NFT Locker address that will receive the NFT and ETH Discount.
+     */
+    function nftLocker() public view virtual returns (address) {
+        return _nftLocker;
+    }
+
+    /**
+     * @dev Returns the beneficiary that will receive the remaining ETH.
+     */
+    function beneficiary() public view virtual returns (address) {
+        return _beneficiary;
+    }
+
+    /**
+     * @dev Returns the time when the NFT can be released, and vesting starts in seconds since Unix epoch (i.e. Unix timestamp).
+     */
+    function vestingStartTime() public view virtual returns (uint256) {
+        return _vestingStartTime;
+    }
+
+    /**
+     * @dev Returns the maximum number of intervals set for vesting.
+     */
+    function maxIntervals() public view virtual returns (uint256) {
+        return _maxIntervals;
+    }
+
+    /**
+     * @dev Returns the set duration of each interval.
+     */
+    function intervalDuration() public view virtual returns (uint256) {
+        return _intervalDuration;
+    }
+
+    /**
+     * @dev Returns duration that NFT has been locked and vesting
+     */
+    function vestedDuration() public view returns (uint256) {
+        return block.timestamp - vestingStartTime();
+    }
+
+    /**
+     * @dev Returns the number of interval that the token has been vested for after the vesting start time.
+     */
+    function intervalsPassed() public view returns (uint256) {
+        return vestedDuration() / intervalDuration();
+    }
+
+    /**
+     * @dev Returns current vesting interval.
+     */
+    function currentInterval() public view returns (uint256) {
+        // Before vesting start time, the interval is 0.
+        if (block.timestamp < vestingStartTime()) {
+            return 0;
+        }
+
+        // After vesting start time, the interval interval count turns to 1.
+        uint256 intervals = 1 + intervalsPassed();
+
+        if (intervals > maxIntervals()) {
+            intervals = maxIntervals();
+        }
+        return intervals;
+    }
+
+    /**
+     * @dev Returns the remaining interval before max vesting interval
+     */
+    function getIntervalsLeft() public view returns (uint256) {
+        return maxIntervals() - currentInterval();
+    }
+
+    /**
+     * @dev Returns discount accrued by the user up until the current interval.
+     *      Discount ratio is the ratio of number of interval passed {currentInterval} to {maxIntervals}.
+     *      Maximum discount ratio is 1.
+     */
+    function getDiscount() public view returns (uint256) {
+        return currentInterval() * (address(this).balance / maxIntervals());
```

Comment:
Add comment on why the need for balance / maxIntervals due to the mathematics issue

Reply:
Add comment here as well